File: src/lib/professions.ts
/**
 * Single source of truth for the FOCUS_AREA label -> profession_slug
 * mapping.
 *
 * ROOT CAUSE this file fixes: `professions.slug` (migrations/001, seeded
 * once and never re-derived) and the `PROFESSION_SLUGS` weight-table keys
 * (scoring/professionWeights.ts) were both hand-written to *abbreviate*
 * multi-word labels by dropping "and" (e.g. "Programming & Tech" ->
 * `programming_tech`). Meanwhile src/server/routes/discover.ts and
 * src/server/routes/intelligence.ts each hand-rolled their own
 * `slugifyProfession(label)` that mechanically lowercases the label and
 * turns "&" into the literal word "and" (`programming_and_tech`, not the
 * `programming_tech` the table actually has). Every Discover request whose
 * profile had a multi-word focus area ("Programming & Tech", "Writing &
 * Translation", "Video & Animation", "Personal Growth & Hobbies")
 * therefore generated a `profession_slug` that no row in `professions`
 * has, and the later `leads` insert (profession_slug references
 * professions(slug)) failed its foreign key check every time.
 *
 * Fix: stop *generating* slugs from labels via string transformation
 * entirely. Slugs are looked up from this fixed table instead, so a
 * profession_slug can only ever be one of the values already seeded into
 * `professions` (migrations/001_opportunity_engine.sql) — it is
 * structurally impossible for this lookup to produce a slug the table
 * doesn't have, because the two lists here ARE what the seed and the
 * scoring weight table are generated from. professionWeights.ts and
 * onboarding.tsx's FOCUS_AREAS both defer to this module now.
 */

export const PROFESSION_SLUGS = [
  "graphic_design",
  "digital_marketing",
  "writing_translation",
  "video_animation",
  "music_audio",
  "programming_tech",
  "data",
  "business",
  "personal_growth_hobbies",
  "photography",
  "finance",
  "end_to_end_project",
] as const;

export type ProfessionSlug = (typeof PROFESSION_SLUGS)[number];

/**
 * Onboarding's FOCUS_AREAS labels, in the exact same order as
 * PROFESSION_SLUGS above — index i of one corresponds to index i of the
 * other. onboarding.tsx zips its icons onto these labels; nothing else
 * should hand-write this label list again.
 */
export const FOCUS_AREA_LABELS = [
  "Graphic Design",
  "Digital Marketing",
  "Writing & Translation",
  "Video & Animation",
  "Music & Audio",
  "Programming & Tech",
  "Data",
  "Business",
  "Personal Growth & Hobbies",
  "Photography",
  "Finance",
  "End-to-End Project",
] as const;

if (FOCUS_AREA_LABELS.length !== PROFESSION_SLUGS.length) {
  // Fails loudly at import time (in every process that loads this module —
  // frontend build, gateway boot, worker boot) instead of silently
  // producing mismatched data if someone edits one list without the other.
  throw new Error("professions.ts: FOCUS_AREA_LABELS and PROFESSION_SLUGS have drifted out of sync");
}

const LABEL_TO_SLUG: ReadonlyMap<string, ProfessionSlug> = new Map(
  FOCUS_AREA_LABELS.map((label, i) => [label, PROFESSION_SLUGS[i]]),
);

/**
 * Resolves a stored `settings.focusArea` label to the profession_slug that
 * actually exists in the `professions` table — or null if the label isn't
 * (or is no longer) one of the fixed FOCUS_AREAS options (e.g. legacy /
 * blank / free-text data). Never invents a slug via string manipulation,
 * so the result is always either a real row in `professions` or null.
 */
export function professionSlugForLabel(label: string | null | undefined): ProfessionSlug | null {
  if (!label) return null;
  return LABEL_TO_SLUG.get(label) ?? null;
}

export function isProfessionSlug(value: string | null | undefined): value is ProfessionSlug {
  return !!value && (PROFESSION_SLUGS as readonly string[]).includes(value);
}

File: mast-lead-engine/opportunity_scoring/professions.py
"""
opportunity_scoring/professions.py
==================================

Canonical profession definitions and normalized weight vectors for opportunity scoring
in Engine 2.0.

Design Rules
------------
- Single Python source of truth for the 12 canonical profession slugs.
- Immutable weight vectors normalized to sum to 1.0 per profession.
- Consumes standard library types only.
"""

from __future__ import annotations

from dataclasses import dataclass

PROFESSION_SLUGS: tuple[str, ...] = (
    "graphic_design",
    "digital_marketing",
    "writing_translation",
    "video_animation",
    "music_audio",
    "programming_tech",
    "data",
    "business",
    "personal_growth_hobbies",
    "photography",
    "finance",
    "end_to_end_project",
)


@dataclass(frozen=True, slots=True)
class WeightVector:
    website: float
    branding: float
    social: float
    growth: float
    newness: float
    tech: float


RAW_WEIGHTS: dict[str, dict[str, float]] = {
    "graphic_design":          {"website": 0.10, "branding": 0.45, "social": 0.20, "growth": 0.10, "newness": 0.15, "tech": 0.03},
    "digital_marketing":       {"website": 0.15, "branding": 0.20, "social": 0.40, "growth": 0.15, "newness": 0.10, "tech": 0.05},
    "writing_translation":     {"website": 0.30, "branding": 0.20, "social": 0.20, "growth": 0.15, "newness": 0.15, "tech": 0.03},
    "video_animation":         {"website": 0.10, "branding": 0.30, "social": 0.35, "growth": 0.10, "newness": 0.15, "tech": 0.03},
    "music_audio":             {"website": 0.15, "branding": 0.25, "social": 0.35, "growth": 0.10, "newness": 0.15, "tech": 0.03},
    "programming_tech":        {"website": 0.45, "branding": 0.08, "social": 0.08, "growth": 0.12, "newness": 0.08, "tech": 0.30},
    "data":                    {"website": 0.30, "branding": 0.08, "social": 0.08, "growth": 0.28, "newness": 0.13, "tech": 0.15},
    "business":                {"website": 0.15, "branding": 0.15, "social": 0.12, "growth": 0.28, "newness": 0.12, "tech": 0.20},
    "personal_growth_hobbies": {"website": 0.18, "branding": 0.28, "social": 0.28, "growth": 0.05, "newness": 0.15, "tech": 0.03},
    "photography":             {"website": 0.10, "branding": 0.35, "social": 0.28, "growth": 0.05, "newness": 0.19, "tech": 0.02},
    "finance":                 {"website": 0.13, "branding": 0.08, "social": 0.08, "growth": 0.35, "newness": 0.22, "tech": 0.10},
    "end_to_end_project":      {"website": 0.18, "branding": 0.18, "social": 0.18, "growth": 0.18, "newness": 0.18, "tech": 0.10},
}


def _normalize_weights(raw: dict[str, dict[str, float]]) -> dict[str, WeightVector]:
    normalized: dict[str, WeightVector] = {}
    for slug, w in raw.items():
        total = sum(w.values())
        normalized[slug] = WeightVector(
            website=w["website"] / total,
            branding=w["branding"] / total,
            social=w["social"] / total,
            growth=w["growth"] / total,
            newness=w["newness"] / total,
            tech=w["tech"] / total,
        )
    return normalized


PROFESSION_WEIGHTS: dict[str, WeightVector] = _normalize_weights(RAW_WEIGHTS)
